### PR TITLE
IL2CPP debug images, fixes and tests

### LINKED ISSUE #897: Filter debug images added to IL2CPP exceptions
Originally discussion & TODO in https://github.com/getsentry/sentry-unity/pull/896#discussion_r925619698_

We should only add debug images that are present in the current stack trace (the rest is useless data on an event). Such filtering is already in place on mobile SDKs.

### PR DESCRIPTION
**Pull request by vaind**
Tested (inspected debug logs) manually on:
* Windows Unity 2020, 2021, 2022
* Linux Unity 2021
* Android Unity 2021
* iOS Unity 2021

closes #897 

--------------------
Outdated:
* before: https://sentry.io/organizations/sentry-sdks/issues/3441819954/events/52c89fbf31014683954a7f7477829da7/?project=5439417
* after: https://sentry.io/organizations/sentry-sdks/issues/3441819954/events/d25ae586cd7c4a768c8564d43b7a0b1a/?project=5439417

partial diff:
image

### CODE REVIEW

**Comment by bruno-garcia** [comment]
Neither before nor after screenshot has line numbers, should it?

**Comment by vaind** [comment]
> > Neither before nor after screenshot has line numbers, should it?
> 
> I see now that I was supposed to enable something... Let me try again

It all seems to be enabled and the exception processor runs, I just don't see anything useful in the event. Maybe it only works on some events? I haven't been involved with this feature so can't tell for sure if I'm testing it wrong (pressing the "Throw null" button in the Example)

Also, I couldn't run on iOS because it just doesn't get invoked (no il2cpp methods for some reason, despite this issue has it checked as working: #895). And macOS because I can't run on Apple silicon with newer Unity (buggy build)

**Comment by Swatinem** [comment]
Meh, GitHub does not allow me to comment on "unrelated" code, so here goes:

You have this `var imageAddress = 0x1000;` with a huge comment above it.
I think you can make it into:

```
// We want to use an address that is definitely outside of any address range used by real libraries.
// Canonical addresses on x64 leave a gap in the middle of the address space, which is unused.
// This is a range of addresses that we should be able to safely use.
uint64 imageAddress = 1 << 63;
```

File: src/Sentry.Unity/Il2CppEventProcessor.cs [hunk 1]
```diff
@@ -32,9 +33,13 @@ public void Process(Exception incomingException, SentryEvent sentryEvent)
             }
             var exceptions = EnumerateChainedExceptions(incomingException);
 
+            var debugImages = (sentryEvent.DebugImages ??= new List<DebugImage>());
+
+            // TODO maybe filter only images needed for the current stack trace instead of adding all of them?
```
[the remark this answers is not in the document]
Comment (vaind) [reply src/Sentry.Unity/Il2CppEventProcessor.cs:38]: https://github.com/getsentry/sentry-unity/issues/897

**Comment by vaind** [comment]
> 

I've applied that change, seems to work fine AFAICT

File: src/Sentry.Unity/Il2CppEventProcessor.cs [hunk 2]
```diff
@@ -91,7 +93,7 @@ public void Process(Exception incomingException, SentryEvent sentryEvent)
                     // The instructions in the stack trace generally have "return addresses"
                     // in them. But for symbolication, we want to symbolicate the address of
                     // the "call instruction", which in almost all cases happens to be
-                    // the instruction right in front of the return address.
+                    // the instruction right in front of the return address .
```
Comment (loewenheim) [inline src/Sentry.Unity/Il2CppEventProcessor.cs:96]: This seems like a typo.

**Review by loewenheim — APPROVED** [review]

File: src/Sentry.Unity/Il2CppEventProcessor.cs [hunk 3]
```diff
@@ -56,47 +54,163 @@ public void Process(Exception incomingException, SentryEvent sentryEvent)
...
+                // We use this when we encounter stack frames with relative addresses.
+                // We want to use an address that is definitely outside of any address range used by real libraries.
+                // Canonical addresses on x64 leave a gap in the middle of the address space, which is unused.
+                // This is a range of addresses that we should be able to safely use.
+                // See https://en.wikipedia.org/wiki/X86-64#Virtual_address_space_details
+                var mainLibOffset = (ulong)1 << 63;
+                DebugImage? mainLibImage = null;
 
+                // TODO do we really want to continue if these two don't match?
+                //      Wouldn't it cause invalid frame info?
```
Comment (Swatinem) [inline src/Sentry.Unity/Il2CppEventProcessor.cs:69]: good point, it probably would, I suspect it would only be on the bottom of the stack though.

File: src/Sentry.Unity/Il2CppEventProcessor.cs [hunk 4]
```diff
@@ -56,47 +54,163 @@ public void Process(Exception incomingException, SentryEvent sentryEvent)
...
+                }
+            }
+
+            sentryEvent.DebugImages ??= new List<DebugImage>();
+            sentryEvent.DebugImages.AddRange(usedImages);
+        }
+
+        // Normalizes Debug Image UUID so that we can compare the ones coming from
+        // native (contains dashes, all lower-case) & what Unity gives us (no dashes, uppercase).
+        // On Linux, the image also has shorter UUID coming from Unity, e.g. 3028cb80b0712541,
```
Comment (Swatinem) [inline src/Sentry.Unity/Il2CppEventProcessor.cs:141]: thats surprising. I thought it would be even longer (20 bytes instead of 16)

File: src/Sentry.Unity/Il2CppEventProcessor.cs [hunk 5]
```diff
@@ -56,47 +54,163 @@ public void Process(Exception incomingException, SentryEvent sentryEvent)
...
+                    "Found debug image '{0}' (CodeId {1} | DebugId: {2}) with addresses between {3:X8} and {4:X8}",
+                    null, image.CodeFile, image.CodeId, image.DebugId, info.StartAddress, info.EndAddress);
+            }
+            return result;
+        });
+
+        private static DebugImage? FindDebugImageContainingAddress(ulong instructionAddress)
+        {
+            var list = DebugImagesSorted.Value;
+            foreach (var info in list)
```
Comment (Swatinem) [inline src/Sentry.Unity/Il2CppEventProcessor.cs:198]: doesn’t C# have a binary search in the stdlib? iterating the whole list for each frame may be expensive

File: src/Sentry.Unity/Il2CppEventProcessor.cs [hunk 6]
```diff
@@ -56,47 +54,163 @@ public void Process(Exception incomingException, SentryEvent sentryEvent)
...
 
                     // The instructions in the stack trace generally have "return addresses"
                     // in them. But for symbolication, we want to symbolicate the address of
                     // the "call instruction", which in almost all cases happens to be
                     // the instruction right in front of the return address.
                     // A good heuristic to use in that case is to just subtract 1.
-                    var instructionAddr = imageAddress + nativeFrame.ToInt64() - 1;
-                    frame.InstructionAddress = $"0x{instructionAddr:X8}";
+                    var instructionAddress = (ulong)nativeFrame.ToInt64() - 1;
+                    var image = FindDebugImageContainingAddress(instructionAddress);
```
Comment (Swatinem) [inline src/Sentry.Unity/Il2CppEventProcessor.cs:86]: this might give you some false positives :-(

In one of the examples I have seen that one of the images had an addr at `0x40_000`, and there were some GameAssembly-relative addresses that just happen to be inside of that :-(

**Review by Swatinem — COMMENTED** [review]
maybe I’m running a bit in circles here, this seems to be a hard problem to solve.
also considering we want to limit the number of images we put into the event payload.

We have two facts of how unity works that can help us along:

1. Unity filters the stack trace so it only outputs stuff thats coming from the GameAssembly
2. We have either absolute addrs, where *all* the addresses are within the addr range that native reports for the GameAssembly, or we have relative addrs where *some* may be outside of that range.

How about:
- We find the *real* GameAssembly as reported by the native SDK, matching by UUID
- If *all* the addresses are inside it, we have absolute addrs, if *some* are outside of it, we have relative addrs.

We need to test for this with *all* addresses at once, not for each one on its own.

File: src/Sentry.Unity/Il2CppEventProcessor.cs (hunk 5 above)
Comment (vaind) [reply src/Sentry.Unity/Il2CppEventProcessor.cs:198]: Changed to binary search, though I think it's a bit less readable

File: src/Sentry.Unity/Il2CppEventProcessor.cs (hunk 6 above)
Comment (vaind) [reply src/Sentry.Unity/Il2CppEventProcessor.cs]: The issue is, now all the code only matches through the package-name code branch... At least with the error examples we have in the sample. Any ideas how to create stack trace with items that would match other debug images? Otherwise it seems kinda pointless to even post them.

Comment (vaind) [reply src/Sentry.Unity/Il2CppEventProcessor.cs]: BTW I've tried adding an example with a callback through native code, but that is still part of GameAssembly. Other option I can think of is using an external library, e.g. sentry, with one of its callbacks, e.g. before_send

**Review by bitsandfoxes — APPROVED** [review]
